File: API/PetCenterModels/DataTransferObjects/Listing/SubDTOs.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Text;
using System.Text.Json.Serialization;
using System.Threading.Tasks;
using PetCenterModels.DBTables;

namespace PetCenterModels.DataTransferObjects
{

    public class ProductListingSubDTO : ISerializableRequestDTO<ProductListing>, IBaseResponseDTO<ProductListing,ProductListingSubDTO>
    {
        public Guid? Id {get; set;}

        public byte[] CurrentVersion { get; set; } = Array.Empty<byte>();

        public List<NoteSubDTO>? Notes {get; set;} = null;

        public Guid ProductId {get; set;} = Guid.Empty;

        public byte PerListing {get; set;} = 1;


        public bool Validate()
        {
            return !(ProductId==Guid.Empty) && (PerListing>0);
        }

        public ProductListing? ToEntity()
        {
            ProductListing output = new();
            output.CurrentVersion=CurrentVersion;
            output.ProductId=ProductId;
            output.PerListing=PerListing;
            return output;   
        }

        public static ProductListingSubDTO? FromEntity(ProductListing? entity)
        {
            if(entity==null){return null;}
            return new ProductListingSubDTO
            {
                Id=entity.Id,
                CurrentVersion=entity.CurrentVersion,
                ProductId=entity.ProductId,
                PerListing=entity.PerListing
            };

        }

    }


    public class AnimalListingSubDTO : ISerializableRequestDTO<AnimalListing>, IBaseResponseDTO<AnimalListing,AnimalListingSubDTO>
    {
        public Guid? Id {get; set;}

        public byte[] CurrentVersion { get; set; } = Array.Empty<byte>();

        public List<NoteSubDTO>? Notes {get; set;} = null;

        public Guid AnimalId {get; set;} = Guid.Empty;


        public bool Validate()
        {
            return !(AnimalId==Guid.Empty);
        }

        public AnimalListing? ToEntity()
        {
            AnimalListing output = new();
            output.CurrentVersion=CurrentVersion;
            output.AnimalId=AnimalId;
            return output;   
        }

        public static AnimalListingSubDTO? FromEntity(AnimalListing? entity)
        {
            if(entity==null){return null;}
            return new AnimalListingSubDTO
            {
                Id=entity.Id,
                CurrentVersion=entity.CurrentVersion,
                AnimalId = entity.AnimalId
            };

        }

    }


    public class MedicalListingSubDTO : ISerializableRequestDTO<MedicalListing>, IBaseResponseDTO<MedicalListing,MedicalListingSubDTO>
    {
        public Guid? Id {get; set;}

        public byte[] CurrentVersion { get; set; } = Array.Empty<byte>();

        public List<NoteSubDTO>? Notes {get; set;} = null;

        public Guid ProcedureId {get; set;} = Guid.Empty;


        public bool Validate()
        {
            return !(ProcedureId==Guid.Empty);
        }

        public MedicalListing? ToEntity()
        {
            MedicalListing output = new();
            output.CurrentVersion=CurrentVersion;
            output.ProcedureId=ProcedureId;
            return output;   
        }

        public static MedicalListingSubDTO? FromEntity(MedicalListing? entity)
        {
            if(entity==null){return null;}
            return new MedicalListingSubDTO
            {
                Id=entity.Id,
                CurrentVersion=entity.CurrentVersion,
                ProcedureId = entity.ProcedureId
            };

        }

    }


    public class CommentResponseSubDTO : IBaseResponseDTO<Comment, CommentResponseSubDTO>
    {
        public Guid? Id {get; set;}

        public byte[] CurrentVersion { get; set; } = Array.Empty<byte>();

        public Guid PosterId {get; set;} = Guid.Empty;

        public string PosterName {get; set;} = string.Empty;

        public string Contents {get; set;} = string.Empty;

        public List<NoteSubDTO>? Notes {get; set;} = null;

        public static CommentResponseSubDTO? FromEntity(Comment? entity)
        {
            if(entity==null){return null;}
            CommentResponseSubDTO output = new CommentResponseSubDTO
            {
                Id=entity.Id,
                CurrentVersion=entity.CurrentVersion,
                PosterId=entity.PosterId,
                PosterName="Anonymous",
                Contents=entity.Message
            };

            if (entity.Poster != null)
            {
                output.PosterName=entity.Poster.UserName;
            }

            return output;
        }

    }

    public class AvailabilityResponseSubDTO : IBaseResponseDTO<Available, AvailabilityResponseSubDTO>
    {
        public Guid? Id {get; set;}

        public byte[] CurrentVersion { get; set; } = Array.Empty<byte>();

        public Guid FacilityId {get; set;}

        public string Contact {get; set;} = string.Empty;

        public string City {get; set;} = string.Empty;

        public string Street {get; set;} = string.Empty;

        public List<NoteSubDTO>? Notes {get; set;} = null;

        public static AvailabilityResponseSubDTO? FromEntity(Available? entity)
        {
            if(entity==null){return null;}
            AvailabilityResponseSubDTO output = new AvailabilityResponseSubDTO
            {
                Id=entity.Id,
                CurrentVersion=entity.CurrentVersion,
                FacilityId=entity.FacilityId,
                Contact="No contact provided.",
                City=entity.RelevantFacility.City="No city provided.",
                Street=entity.RelevantFacility.Street="No street provided."

            };

            if (entity.RelevantFacility != null)
            {
                if (entity.RelevantFacility.OwningFranchise != null)
                {
                    output.Contact=entity.RelevantFacility.OwningFranchise.Contact;
                }

                output.Contact=entity.RelevantFacility.Contact??output.Contact;
                
                output.City=entity.RelevantFacility.City??output.City;
                output.Street=entity.RelevantFacility.Street??output.Street;
            }

            return output;

        }

    }


    public class DiscountResponseSubDTO : IBaseResponseDTO<Discount, DiscountResponseSubDTO>
    {
        public Guid? Id {get; set;}

        public byte[] CurrentVersion { get; set; } = Array.Empty<byte>();

        public int Percentage {get; set;} = 0;

        public DateTime Expiry {get; set;} = DateTime.UtcNow;

        public List<NoteSubDTO>? Notes {get; set;} = null;

        public static DiscountResponseSubDTO? FromEntity(Discount? entity)
        {
            if(entity==null){return null;}
            return new DiscountResponseSubDTO
            {
                Id=entity.Id,
                CurrentVersion=entity.CurrentVersion,
                Percentage=entity.PercentDiscount,
                Expiry=entity.Expiry

            };

        }

    }


    public class ReportResponseSubDTO : IBaseResponseDTO<Report, ReportResponseSubDTO>
    {
        public Guid? Id {get; set;}

        public byte[] CurrentVersion { get; set; } = Array.Empty<byte>();

        public string Reason {get; set;} = string.Empty;

        public Guid ReporterId {get; set;} = Guid.Empty;

        public Guid ListingId {get; set;} = Guid.Empty;

        public Guid? CommentId {get; set;} = null;

        public DateTime Expiry {get; set;} = DateTime.UtcNow;

        public List<NoteSubDTO>? Notes {get; set;} = null;

        public static ReportResponseSubDTO? FromEntity(Report? entity)
        {
            if(entity==null){return null;}
            return new ReportResponseSubDTO
            {
                Id=entity.Id,
                CurrentVersion=entity.CurrentVersion,
                Reason=entity.Reason,
                ReporterId=entity.ReporterId,
                ListingId=entity.ListingId,
                CommentId=entity.CommentId,
                Expiry=entity.Expiry

            };

        }

    }

}

File: Frontend/pet_center_app/lib/models/data_transfer/listing/sub_dtos.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:pet_center_app/models/data_transfer/note_sub_dto.dart';
part 'sub_dtos.g.dart';

@JsonSerializable(explicitToJson: true)
class ProductListingSubDTO {
  String? id;
  String currentVersion;
  List<NoteSubDTO>? notes;
  String productId;
  int perListing;

  ProductListingSubDTO({
    this.id,
    this.currentVersion = '',
    this.notes,
    this.productId = '',
    this.perListing = 1,
  });

  factory ProductListingSubDTO.fromJson(Map<String, dynamic> json) =>
      _$ProductListingSubDTOFromJson(json);

  Map<String, dynamic> toJson() => _$ProductListingSubDTOToJson(this);

  ProductListingSubDTO copy() => ProductListingSubDTO(
    id: id,
    currentVersion: currentVersion,
    notes: notes?.map((n) => n.copy()).toList(),
    productId: productId,
    perListing: perListing,
  );
}

@JsonSerializable(explicitToJson: true)
class AnimalListingSubDTO {
  String? id;
  String currentVersion;
  List<NoteSubDTO>? notes;
  String animalId;

  AnimalListingSubDTO({
    this.id,
    this.currentVersion = '',
    this.notes,
    this.animalId = '',
  });

  factory AnimalListingSubDTO.fromJson(Map<String, dynamic> json) =>
      _$AnimalListingSubDTOFromJson(json);

  Map<String, dynamic> toJson() => _$AnimalListingSubDTOToJson(this);

  AnimalListingSubDTO copy() => AnimalListingSubDTO(
    id: id,
    currentVersion: currentVersion,
    notes: notes?.map((n) => n.copy()).toList(),
    animalId: animalId,
  );
}

@JsonSerializable(explicitToJson: true)
class MedicalListingSubDTO {
  String? id;
  String currentVersion;
  List<NoteSubDTO>? notes;
  String procedureId;

  MedicalListingSubDTO({
    this.id,
    this.currentVersion = '',
    this.notes,
    this.procedureId = '',
  });

  factory MedicalListingSubDTO.fromJson(Map<String, dynamic> json) =>
      _$MedicalListingSubDTOFromJson(json);

  Map<String, dynamic> toJson() => _$MedicalListingSubDTOToJson(this);

  MedicalListingSubDTO copy() => MedicalListingSubDTO(
    id: id,
    currentVersion: currentVersion,
    notes: notes?.map((n) => n.copy()).toList(),
    procedureId: procedureId,
  );
}

@JsonSerializable(explicitToJson: true)
class CommentResponseSubDTO {
  String? id;
  String currentVersion;
  String listingId;
  String posterId;
  String posterName;
  String contents;
  List<NoteSubDTO>? notes;
  DateTime lastEditDate;

  CommentResponseSubDTO({
    this.id,
    this.currentVersion = '',
    this.listingId = '',
    this.posterId = '',
    this.posterName = '',
    this.contents = '',
    this.notes,
    DateTime? lastEditDate,
  }) : lastEditDate = lastEditDate ?? DateTime.now().toUtc();

  factory CommentResponseSubDTO.fromJson(Map<String, dynamic> json) =>
      _$CommentResponseSubDTOFromJson(json);

  Map<String, dynamic> toJson() => _$CommentResponseSubDTOToJson(this);

  CommentResponseSubDTO copy() => CommentResponseSubDTO(
    id: id,
    currentVersion: currentVersion,
    listingId: listingId,
    posterId: posterId,
    posterName: posterName,
    contents: contents,
    notes: notes?.map((n) => n.copy()).toList(),
  );
}

@JsonSerializable(explicitToJson: true)
class AvailabilityResponseSubDTO {
  String? id;
  String currentVersion;
  String facilityId;
  String contact;
  String city;
  String street;
  List<NoteSubDTO>? notes;

  AvailabilityResponseSubDTO({
    this.id,
    this.currentVersion = '',
    this.facilityId = '',
    this.contact = '',
    this.city = '',
    this.street = '',
    this.notes,
  });

  factory AvailabilityResponseSubDTO.fromJson(Map<String, dynamic> json) =>
      _$AvailabilityResponseSubDTOFromJson(json);

  Map<String, dynamic> toJson() => _$AvailabilityResponseSubDTOToJson(this);

  AvailabilityResponseSubDTO copy() => AvailabilityResponseSubDTO(
    id: id,
    currentVersion: currentVersion,
    facilityId: facilityId,
    contact: contact,
    city: city,
    street: street,
    notes: notes?.map((n) => n.copy()).toList(),
  );
}

@JsonSerializable(explicitToJson: true)
class DiscountResponseSubDTO {
  String? id;
  String currentVersion;
  int percentage;
  DateTime expiry;
  List<NoteSubDTO>? notes;

  DiscountResponseSubDTO({
    this.id,
    this.currentVersion = '',
    this.percentage = 0,
    DateTime? expiry,
    this.notes,
  }) : expiry = expiry ?? DateTime.now().toUtc();

  factory DiscountResponseSubDTO.fromJson(Map<String, dynamic> json) =>
      _$DiscountResponseSubDTOFromJson(json);

  Map<String, dynamic> toJson() => _$DiscountResponseSubDTOToJson(this);

  DiscountResponseSubDTO copy() => DiscountResponseSubDTO(
    id: id,
    currentVersion: currentVersion,
    percentage: percentage,
    expiry: expiry,
    notes: notes?.map((n) => n.copy()).toList(),
  );
}

@JsonSerializable(explicitToJson: true)
class ReportResponseSubDTO {
  String? id;
  String currentVersion;
  String reason;
  String reporterId;
  String listingId;
  String? commentId;
  DateTime expiry;
  List<NoteSubDTO>? notes;
  DateTime datePosted;

  ReportResponseSubDTO({
    this.id,
    this.currentVersion = '',
    this.reason = '',
    this.reporterId = '',
    this.listingId = '',
    this.commentId,
    DateTime? expiry,
    this.notes,
    DateTime? datePosted,
  }) : expiry = expiry ?? DateTime.now().toUtc(),
       datePosted = datePosted ?? DateTime.now().toUtc();

  factory ReportResponseSubDTO.fromJson(Map<String, dynamic> json) =>
      _$ReportResponseSubDTOFromJson(json);

  Map<String, dynamic> toJson() => _$ReportResponseSubDTOToJson(this);

  ReportResponseSubDTO copy() => ReportResponseSubDTO(
    id: id,
    currentVersion: currentVersion,
    reason: reason,
    reporterId: reporterId,
    listingId: listingId,
    commentId: commentId,
    expiry: expiry,
    notes: notes?.map((n) => n.copy()).toList(),
  );
}
